[757] https tls cert chains

aklyuchev opened the pull request:
------------------------
Note: by creating a PR or an issue you automatically agree to the CLA. See [CONTRIBUTING.md](https://github.com/userver-framework/userver/blob/develop/CONTRIBUTING.md). Feel free to remove this note, the agreement holds.

aklyuchev86 committed:
added certificate chains for http tls connection

aklyuchev86 committed:
fix naming

fdr400 commented:
It it possible to add tests for it?

aklyuchev commented:
> It it possible to add tests for it?

tests for what?
for openssl function which is used here?
i don't think it make sense

apolukhin commented on core/src/server/component.cpp:
```diff
@@ -128,6 +128,9 @@ additionalProperties: false
                         description: path to TLS certificate
+                    cert-chain:
+                        type: string
+                        description: path to TLS certificates chain
```
Please add `defaultDescription: ''` to highlight that the valie is optional

aklyuchev replied:
added

apolukhin commented on core/src/server/component.cpp:
```diff
@@ -128,6 +128,9 @@ additionalProperties: false
                     cert:
                         type: string
                         description: path to TLS certificate
+                    cert-chain:
```
Please, also add the description to https://github.com/userver-framework/userver/blob/bb37e2121cc43ea9317d089435c6ac8e5c41e51b/core/include/userver/server/component.hpp#L66

aklyuchev replied:
added

apolukhin commented on core/src/engine/io/tls_wrapper.cpp:
```diff
@@ -540,6 +541,18 @@ TlsWrapper TlsWrapper::StartTlsServer(
+      }
+  }
+  else {
+      if (1 != SSL_CTX_use_certificate(ssl_ctx.get(), cert.GetNative())) {
```
Thos block is duplicated down below. Looks like the latter is redundant

aklyuchev replied:
removed

apolukhin commented:
> > It it possible to add tests for it?
> 
> tests for what? for openssl function which is used here? i don't think it make sense

It makes sense from the point of documentation. OpenSSL is overcomplicated, especially for the beginners. A test with some hints on how to create tls chains and how the file is named would be helpful. It also makes sense as a smoke test - the openssl function may require some other functions to be called before setting the chain... a test would prevent breaking the functionality during refactorings.

Please, add a smoke test to https://github.com/userver-framework/userver/blob/develop/core/src/engine/io/tls_wrapper_test.cpp

aklyuchev86 committed:
[Tb-926] http cert chains cleanup

apolukhin reviewed (changes requested):
it took some time for me to realize the drawback of this PR. `SSL_CTX_use_certificate_chain_file` does blocking IO operations (reads data from disk). We try hard to do the blocking operations not in the main task processor for CPU bound tasks. So the file reading should **not** be done in `TlsWrapper::StartTlsServer`.

It looks like the certificates could be extracted from chain file https://unix.stackexchange.com/questions/368123/how-to-extract-the-root-ca-and-subordinate-ca-from-a-certificate-chain-in-linux . After that the certificates could be passed via `cert`, `key` and `extra_cert_authorities` parameters of the `TlsWrapper::StartTlsServer` function.

The PR could be changed to do the following:
* Make a function `crypto::blocking::LoadFromChainFile` that returns crypto::Certificate's
* Call that function in core/src/server/net/listener_config.cpp and the certificates would make their way to core/src/server/net/listener_impl.cpp

aklyuchev86 committed:
use common logic for tls cert chain

aklyuchev committed:
Merge branch 'develop' into us757-https-tls-chains_http_cert

aklyuchev86 committed:
cleanup

aklyuchev86 committed:
fix compile

aklyuchev86 committed:
test chains

apolukhin commented on core/include/userver/server/component.hpp:
```diff
@@ -63,7 +63,7 @@ namespace components {
 /// backlog | max count of new connections pending acceptance | 1024
 /// tls.ca | paths to TLS CAs for client authentication | -
-/// tls.cert | path to TLS server certificate | -
+/// tls.cert | path to TLS server certificate chain | -
```
How about `path to TLS server certificate or certificate chain`

aklyuchev replied:
ok

apolukhin commented on universal/include/userver/crypto/certificate.hpp:
```diff
@@ -44,6 +45,15 @@ class Certificate {
     std::shared_ptr<NativeType> cert_;
 };
 
+using CertificatesChain = std::list<Certificate>;
```
Why not std::vector<Certificate> ?

aklyuchev replied:
number of certificates in chain is not known in advance

apolukhin commented on universal/src/crypto/certificate.cpp:
```diff
@@ -55,6 +55,29 @@ Certificate Certificate::LoadFromString(std::string_view certificate) {
+CertificatesChain LoadCertficatesChainFromString(std::string_view chain)
+{
+    CertificatesChain certificates;
+    const std::string beginMarker = "-----BEGIN CERTIFICATE-----";
```
constexpr std::string_view

apolukhin commented on universal/src/crypto/certificate.cpp:
```diff
@@ -55,6 +55,29 @@ Certificate Certificate::LoadFromString(std::string_view certificate) {
+{
+    CertificatesChain certificates;
+    const std::string beginMarker = "-----BEGIN CERTIFICATE-----";
+    const std::string endMarker = "-----END CERTIFICATE-----";
```
constexpr std::string_view

apolukhin commented on universal/include/userver/crypto/certificate.hpp:
```diff
@@ -44,6 +45,15 @@ class Certificate {
+/// list of 'Certificate's.
+///
+/// @throw crypto::KeyParseError if failed to load the certificate.
+CertificatesChain LoadCertficatesChainFromString(std::string_view certificatesChain);
```
`std::string_view chain` to suppress annoying warnings about missmatch of parameter names

apolukhin reviewed (changes requested):
Very good work! Thank you very much

Added a few nitpocks

aklyuchev86 committed:
review comments

aklyuchev commented:
addressed comments and created for another MR for sure
https://github.com/userver-framework/userver/pull/823
for checking CI
current might be closed most probably

apolukhin commented:
@aklyuchev many thanks for the PR!

It is a good thing that you've added the test. Our debug build noted an issue with certificate ownership, `X509_up_ref(cert_it->GetNative());` was missing. Merged with fix in https://github.com/userver-framework/userver/commit/f9dd97a5a1485184207a6d1cf3b1af6cd0441d71
